pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,159 @@
+    gates of a quantum circuit without the need of calculating the gradient of the function. The
+    algorithm works by updating the parameters :math:`\theta = \theta_1, \dots, \theta_D` and gate
+    choices :math:`R = R_1,\dots,R_D` one at a time according to a closed-form expression for the
+    optimal value of the :math:`d^{th}` parameter :math:`\theta^*_d` when the other parameters and
+    gate choices are fixed:
+
+    .. math:: \theta^*_d = \underset{\theta_d}{\text{argmin}}\left<H\right>_{\theta_d}
+              = -\frac{\pi}{2} - \text{arctan2}\left(2\left<H\right>_{\theta_d=0}
+              - \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2},
+              \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2}\right),
+
+    where :math:`\left<H\right>_{\theta_d}` is the expectation value of the objective function
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_.
+
+    Keyword Args:
+        possible_generators (list[~.Operation]): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+
+    **Example:**
+
+    Initialize the optimizer, set the initial values of ``x``, choose the initial generators to
+    start with, and set the number of steps to optimize over.
+
+    >>> opt = qml.optimize.RotoselectOptimizer()
+    >>> x = [0.3, 0.7]
+    >>> generators = [qml.RX, qml.RY]
+    >>> n_steps = 1000
+
+    Define a cost function based on a PennyLane circuit
+
+    >>> def cost(x, generators):
+    >>>     X_1, Y_2 = circuit(x, generators=generators)
+    >>>     return 0.2 * X_1 + 0.5 * Y_2
+
+    Run the optimization step-by-step for ``n_steps`` steps.
+
+    >>> cost_rotosel = []
+    >>> for _ in range(n_steps):
+    >>>     cost_rotosel.append(cost(x, generators))
+    >>>     x, generators = opt.step(cost, x, generators)
+
+    The optimized values for x should now be stored in ``x`` together with the optimal gates for
+    the circuit, while steps-vs-cost can be seen by plotting ``cost_rotosel``.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, **kwargs):
+        self.possible_generators = kwargs.get("possible_generators", [qml.RX, qml.RY, qml.RZ])
+
+    def step(self, objective_fn, x, generators):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization. It should take two lists
+                of the values ``x`` and the gates ``generators`` as inputs, and return a single value.
```

Comment:
Small thing here (not too crucial): would it perhaps be worth defining `x` as a one-dimensional `ndarray` from the start? In the docstring `x` is referred to as a list and later there is a conversion to `ndarray`, this made me think about this, but there might be something I'm missing.

Reply:
Changed all the mentions of ``x`` as a ``list`` to ``array``.